src/controller/DeviceControllerInteractionModelDelegate.h

```diff
@@ -0,0 +1,86 @@
+{
+public:
+    CHIP_ERROR CommandResponseStatus(const app::CommandSender * apCommandSender,
+                                     const Protocols::SecureChannel::GeneralStatusCode aGeneralCode, const uint32_t aProtocolId,
+                                     const uint16_t aProtocolCode, chip::EndpointId aEndpointId, const chip::ClusterId aClusterId,
+                                     chip::CommandId aCommandId, uint8_t aCommandIndex) override;
+
+    CHIP_ERROR CommandResponseProtocolError(const app::CommandSender * apCommandSender, uint8_t aCommandIndex) override;
+
+    CHIP_ERROR CommandResponseError(const app::CommandSender * apCommandSender, CHIP_ERROR aError) override;
+
+    CHIP_ERROR CommandResponseProcessed(const app::CommandSender * apCommandSender) override;
+
+    void OnReportData(const app::ReadClient * apReadClient, const app::ClusterInfo & aPath, TLV::TLVReader * apData,
+                      Protocols::InteractionModel::ProtocolCode status) override;
+    CHIP_ERROR ReadError(const app::ReadClient * apReadClient, CHIP_ERROR aError) override;
+
+    CHIP_ERROR WriteResponseStatus(const app::WriteClient * apWriteClient,
+                                   const Protocols::SecureChannel::GeneralStatusCode aGeneralCode, const uint32_t aProtocolId,
+                                   const uint16_t aProtocolCode, app::AttributePathParams & aAttributePathParams,
+                                   uint8_t aCommandIndex) override;
+
+    CHIP_ERROR WriteResponseProtocolError(const app::WriteClient * apWriteClient, uint8_t aAttributeIndex) override;
+
+    CHIP_ERROR WriteResponseError(const app::WriteClient * apWriteClient, CHIP_ERROR aError) override;
+
+    CHIP_ERROR SubscribeResponseProcessed(const app::ReadClient * apSubscribeClient) override;
+
+    CHIP_ERROR ReadDone(const app::ReadClient * apReadClient) override;
+
+    // TODO: FreeAttributePathParam and AllocateAttributePathParam are used by CHIPDevice.cpp for getting a long-live attribute path
+    // object.
+    void FreeAttributePathParam(uint64_t applicationId)
+    {
+        for (auto & item : mAttributePathTransactionMapPool)
+        {
+            if (item.ApplicationId == applicationId)
+            {
+                item.ApplicationId = UINT64_MAX;
+            }
+        }
+    }
+
+    // TODO: We only support allocating one path, should support multiple path later.
+    app::AttributePathParams * AllocateAttributePathParam(size_t n, uint64_t applicationId)
+    {
+        if (n > 1)
+        {
+            return nullptr;
+        }
+        for (auto & item : mAttributePathTransactionMapPool)
+        {
+            if (item.ApplicationId == UINT64_MAX)
+            {
+                item.ApplicationId = applicationId;
+                return &item.Params;
+            }
+        }
+        return nullptr;
+    }
+
+private:
+    struct AttributePathTransactionMap
+    {
+        uint64_t ApplicationId = UINT64_MAX;
```

Comment:
Why do we need a map keyed on magic numerical Application IDs?  This seems like it will just force extra bookkeeping in the app.  Each subscription is a distinct instance.  Why can't callbacks just be passed a void * context?

Reply:
Maybe I'm misunderstanding.  Is this literally the SubscriptionId from the spec?  If so, that's fine I suppose.  And I presume then that we are using uint64 instead of int32 to accommodate the magic UINT64_MAX ID.

If that is so, I suggest a couple things: (1) rename this to SubscriptionId if that's what it is; (2) add a constexpr for the UINT64_MAX special value so it's clear what that means; (3) make external APIs use uint32 to prevent nonsensical Subscription IDs from being passed in or out of this class.

Reply:
Also, though not changed in this PR, app/MessageDef/ReportData.h has SubscriptionId as a uint64.  I thought for sure subscription IDs were 32bit.  Am I missing something?

Reply:
in latest spec, refreshed on this monday, this becomes 32bit, I would follow-up it immediately after this pr is landed on this ticket, https://github.com/project-chip/connectedhomeip/issues/9520

Reply:
The case is we don't want to expose the underlying objects to the user for memory safety. (i.e. the other commands will use a handle wrapper when possible, and for read / subscriptions, as the input is simplier, we can call the im wrapper directly without using handle)

Also the pointer can be unreliable for classes  inheritance relations (since a pointer pointing to base class and child class  can have different values),.

As all callbacks in interaction model delegate has a pointer to the objects, users CAN use them directly without using the appid.